File: java/runner/java/com/senzing/runner/InstallLocations.java
package com.senzing.runner;

import javax.json.JsonObject;
import java.io.File;
import java.io.StringWriter;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.List;

import static com.senzing.runner.Utilities.*;

/**
 * Describes the directories on disk used to find the Senzing product
 * installation and the support directories.
 */
public class InstallLocations {
    /**
     * The installation location.
     */
    private File installDir;

    /**
     * The location of the configuration files for the config directory.
     */
    private File configDir;

    /**
     * The location of the resource files for the resource directory.
     */
    private File resourceDir;

    /**
     * The location of the support files for the support directory.
     */
    private File supportDir;

    /**
     * The location of the template files for the template directory.
     */
    private File templatesDir;

    /**
     * Indicates if the installation direction is from a development build.
     */
    private boolean devBuild = false;

    /**
     * Default constructor.
     */
    private InstallLocations() {
        this.installDir = null;
        this.configDir = null;
        this.resourceDir = null;
        this.supportDir = null;
        this.templatesDir = null;
        this.devBuild = false;
    }

    /**
     * Gets the primary installation directory.
     *
     * @return The primary installation directory.
     */
    public File getInstallDirectory() {
        return this.installDir;
    }

    /**
     * Gets the configuration directory.
     *
     * @return The configuration directory.
     */
    public File getConfigDirectory() {
        return this.configDir;
    }

    /**
     * Gets the resource directory.
     *
     * @return The resource directory.
     */
    public File getResourceDirectory() {
        return this.resourceDir;
    }

    /**
     * Gets the support directory.
     *
     * @return The support directory.
     */
    public File getSupportDirectory() {
        return this.supportDir;
    }

    /**
     * Gets the templates directory.
     *
     * @return The templates directory.
     */
    public File getTemplatesDirectory() {
        return this.templatesDir;
    }

    /**
     * Checks if the installation is actually a development build.
     * 
     * @return <code>true</code> if this installation represents a development
     *         build, otherwise <code>false</code>.
     */
    public boolean isDevelopmentBuild() {
        return this.devBuild;
    }

    /**
     * Produces a {@link String} describing this instance.
     * 
     * @return A {@link String} describing this instance.
     */
    public String toString() {
        StringWriter sw = new StringWriter();
        PrintWriter pw = new PrintWriter(sw);

        pw.println();
        pw.println("--------------------------------------------------");
        pw.println("installDirectory   : " + this.getInstallDirectory());
        pw.println("configDirectory    : " + this.getConfigDirectory());
        pw.println("supportDirectory   : " + this.getSupportDirectory());
        pw.println("resourceDirectory  : " + this.getResourceDirectory());
        pw.println("templatesDirectory : " + this.getTemplatesDirectory());
        pw.println("developmentBuild   : " + this.isDevelopmentBuild());

        return sw.toString();
    }

    /**
     * Finds the install directories and returns the {@link InstallLocations}
     * instance describing those locations.
     *
     * @param senzingDir The optional Senzing installation directory if one has
     *                   been provided, <code>null</code> otherwise.
     * 
     * @return The {@link InstallLocations} instance describing the install
     *         locations.
     */
    public static InstallLocations findLocations() {
        final String osName = System.getProperty("os.name");

        boolean windows = false;
        boolean macOS = false;

        String lowerOSName = osName.toLowerCase().trim();
        if (lowerOSName.startsWith("windows")) {
            windows = true;
        } else if (lowerOSName.startsWith("mac")
                   || lowerOSName.indexOf("darwin") >= 0)
        {
            macOS = true;
        }

        File installDir = null;
        File configDir = null;
        File resourceDir = null;
        File supportDir = null;
        File templatesDir = null;
        try {
            String defaultInstallPath;
            String defaultConfigPath = null;
            String defaultSupportPath = null;

            if (windows) {
                defaultInstallPath = "C:\\Program Files\\Senzing\\er";
                defaultSupportPath = "C:\\Program Files\\Senzing\\er\\data";
            } else if (macOS) {
                defaultInstallPath = "/opt/senzing/er";
                defaultSupportPath = "/opt/senzing/er/data";
            } else {
                defaultInstallPath = "/opt/senzing/er";
                defaultConfigPath  = "/etc/opt/senzing";
                defaultSupportPath = "/opt/senzing/data";
            }

            // set the install path if one has been provided
            String installPath = System.getProperty("senzing.install.dir");
            String configPath = System.getProperty("senzing.config.dir");
            String supportPath = System.getProperty("senzing.support.dir");
            String resourcePath = System.getProperty("senzing.resource.dir");

            // try environment variables if system properties don't work
            if (installPath == null || installPath.trim().length() == 0) {
                installPath = System.getenv("SENZING_DIR");
            }
            if (configPath == null || configPath.trim().length() == 0) {
                configPath = System.getenv("SENZING_ETC_DIR");
            }
            if (supportPath == null || supportPath.trim().length() == 0) {
                supportPath = System.getenv("SENZING_DATA_DIR");
            }

            // normalize empty strings as null
            if (installPath != null && installPath.trim().length() == 0) {
                installPath = null;
            }
            if (configPath != null && configPath.trim().length() == 0) {
                configPath = null;
            }
            if (supportPath != null && supportPath.trim().length() == 0) {
                supportPath = null;
            }
            if (resourcePath != null && resourcePath.trim().length() == 0) {
                resourcePath = null;
            }

            // check the senzing directory
            installDir = new File(installPath == null ? defaultInstallPath : installPath);
            if (!installDir.exists()) {
                System.err.println("Could not find Senzing installation directory:");
                System.err.println("     " + installDir);
                System.err.println();
                if (installPath != null) {
                    System.err.println(
                        "Check the -Dsenzing.install.dir=[path] command line option.");
                } else {
                    System.err.println(
                        "Use the -Dsenzing.install.dir=[path] command line option to "
                        + "specify a path");
                }

                return null;
            }

            // normalize the senzing directory
            String dirName = installDir.getName();
            if (installDir.isDirectory() && !dirName.equalsIgnoreCase("er")
                && dirName.equalsIgnoreCase("senzing"))
            {
                // for windows or linux allow the "Senzing" dir as well
                installDir = new File(installDir, "er");
            }

            if (!installDir.isDirectory()) {
                System.err.println("Senzing installation directory appears invalid:");
                System.err.println("     " + installDir);
                System.err.println();
                if (installPath != null) {
                    System.err.println(
                        "Check the -Dsenzing.install.dir=[path] command line option.");
                } else {
                    System.err.println(
                        "Use the -Dsenzing.install.dir=[path] command line option to "
                        + "specify a path");
                }

                return null;
            }

            // check if an explicit support path has been specified
            if (supportPath == null || supportPath.trim().length() == 0) {
                // check if using a dev build
                if ("dist".equals(installDir.getName())) {
                    // use the "data" sub-directory of the dev build
                    supportDir = new File(installDir, "data");
                } else {
                    // no explicit path, try the default support path
                    supportDir = new File(defaultSupportPath);
                }

            } else {
                // use the specified explicit path
                supportDir = new File(supportPath);
            }

            if (!supportDir.exists()) {
                System.err.println("The support directory does not exist:");
                System.err.println("         " + supportDir);
                if (supportPath != null) {
                    System.err.println("Check the -Dsenzing.support.dir=[path] command line option.");
                } else {
                    System.err.println("Use the -Dsenzing.support.dir=[path] command line option to "
                        + "specify a path");
                }

                throw new IllegalStateException("The support directory does not exist: " + supportDir);
            }

            if (!supportDir.isDirectory()) {
                System.err.println("The support directory is invalid:");
                System.err.println("         " + supportDir);
                if (supportPath != null) {
                    System.err.println("Check the -Dsenzing.support.dir=[path] command line option.");
                } else {
                    System.err.println("Use the -Dsenzing.support.dir=[path] command line option to "
                        + "specify a path");
                }
                throw new IllegalStateException("The support directory is invalid: " + supportDir);

            }

            // check the config path
            if (configPath != null) {
                configDir = new File(configPath);
            }

            // check for a dev build installation
            if (configDir == null && installDir != null && "dist".equals(installDir.getName())) {
                configDir = new File(installDir, "data");
            }

            // if still null and there is a default, then use it
            if (configDir == null && defaultConfigPath != null) {
                configDir = new File(defaultConfigPath);
                if (!configDir.exists()) {
                    configDir = null;
                }
            }

            // if still null, try to use the install's etc directory
            if (configDir == null && installDir != null) {
                configDir = new File(installDir, "etc");
                if (!configDir.exists()) {
                    configDir = null;
                }                
            }

            if (configPath != null && !configDir.exists()) {
                System.err.println(
                    "The -Dsenzing.config.dir=[path] option specifies a path that does not exist:");
                System.err.println("         " + configPath);

                throw new IllegalStateException("Explicit config path does not exist: " + configPath);
            }
            if (configDir != null && configDir.exists()) {
                if (!configDir.isDirectory()) {
                    System.err.println(
                        "The -Dsenzing.config.dir=[path] option specifies a file, not a directory:");
                    System.err.println("         " + configPath);

                    throw new IllegalStateException(
                        "Explicit config path is not directory: " + configPath);
                }

                String[] requiredFiles = { "cfgVariant.json" };
                List<String> missingFiles = new ArrayList<>(requiredFiles.length);

                for (String fileName : requiredFiles) {
                    File configFile = new File(configDir, fileName);
                    File supportFile = new File(supportDir, fileName);
                    if (!configFile.exists() && !supportFile.exists()) {
                        missingFiles.add(fileName);
                    }
                }
                if (missingFiles.size() > 0 && configPath != null) {
                    System.err.println(
                        "The -Dsenzing.config.dir=[path] option specifies an invalid config directory:");
                    for (String missing : missingFiles) {
                        System.err.println("         " + missing + " was not found");
                    }
                    throw new IllegalStateException(
                            "Explicit config path missing required files: " + missingFiles);
                }
            }

            // now determine the resource path
            resourceDir = (resourcePath == null) ? null : new File(resourcePath);
            if (resourceDir == null) {
                resourceDir = new File(installDir, "resources");
                if (!resourceDir.exists())
                    resourceDir = null;
            }

            if (resourceDir != null && resourceDir.exists() && resourceDir.isDirectory()) {
                templatesDir = new File(resourceDir, "templates");
            }

            if (resourcePath != null) {
                if (!resourceDir.exists()) {
                    System.err.println(
                        "The -Dsenzing.resource.dir=[path] option specifies a path that does not exist:");
                    System.err.println("         " + resourcePath);

                    throw new IllegalStateException(
                        "Explicit resource path does not exist: " + resourcePath);
                }

                if (!resourceDir.isDirectory() || !templatesDir.exists() || !templatesDir.isDirectory()) {
                    System.err.println(
                            "The -Dsenzing.resource.dir=[path] option specifies an invalid "
                            + "resource directory:");
                    System.err.println("         " + resourcePath);

                    throw new IllegalStateException(
                        "Explicit resource path is not valid: " + resourcePath);
                }

            } else if (!resourceDir.exists() || !resourceDir.isDirectory() || !templatesDir.exists()
                    || !templatesDir.isDirectory()) {
                resourceDir = null;
                templatesDir = null;
            }

            // construct and initialize the result
            InstallLocations result = new InstallLocations();
            result.installDir = installDir;
            result.configDir = configDir;
            result.supportDir = supportDir;
            result.resourceDir = resourceDir;
            result.templatesDir = templatesDir;
            result.devBuild = ("dist".equals(installDir.getName()));

            // return the result
            return result;

        } catch (RuntimeException e) {
            e.printStackTrace();
            throw e;

        } catch (Exception e) {
            e.printStackTrace();
            throw new RuntimeException(e);
        }
    }
}

File: csharp/runner/SnippetRunner/InstallLocations.cs
using System.Text;
using System.Text.Json;
using System.Text.RegularExpressions;

using static System.StringComparison;

/// <summary>
/// Describes the directories on disk used to find the Senzing product
/// installation and the support directories.
/// </summary>
public class InstallLocations
{
    /// <summary>
    /// UTF8 encoding constant.
    /// </summary>
    private static readonly Encoding UTF8 = new UTF8Encoding();

    /// <summary>
    /// The installation location.
    /// </summary>
    private DirectoryInfo? installDir;

    /// <summary>
    /// The location of the configuration files for the config directory.
    /// </summary>
    private DirectoryInfo? configDir;

    /// <summary>
    /// The location of the resource files for the resource directory.
    /// </summary>
    private DirectoryInfo? resourceDir;

    /// <summary>
    /// The location of the support files for the support directory.
    /// </summary>
    private DirectoryInfo? supportDir;

    /// <summary>
    /// The location of the template files for the template directory.
    /// </summary>
    private DirectoryInfo? templatesDir;

    /// <summary>
    /// Indicates if the installation direction is from a development build.
    /// </summary>
    private bool devBuild;

    /// <summary>
    /// Default constructor.
    /// </summary>
    private InstallLocations()
    {
        this.installDir = null;
        this.configDir = null;
        this.resourceDir = null;
        this.supportDir = null;
        this.templatesDir = null;
        this.devBuild = false;
    }

    /// <summary>
    /// Gets the primary installation directory.
    /// </summary>
    /// 
    /// <returns>The primary installation directory.</returns>
    public DirectoryInfo? InstallDirectory
    {
        get
        {
            return this.installDir;
        }
    }

    /// <summary>
    /// Gets the configuration directory.
    /// </summary>
    /// 
    /// <returns>The configuration directory.</returns>
    public DirectoryInfo? ConfigDirectory
    {
        get
        {
            return this.configDir;
        }
    }

    /// <summary>
    /// Gets the resource directory.
    /// </summary>
    /// 
    /// <returns>The resource directory.</returns>
    public DirectoryInfo? ResourceDirectory
    {
        get
        {
            return this.resourceDir;
        }
    }

    /// <summary>
    /// Gets the support directory.
    /// </summary>
    /// 
    /// <returns>The support directory.</returns>
    public DirectoryInfo? SupportDirectory
    {
        get
        {
            return this.supportDir;
        }
    }

    /// <summary>
    /// Gets the templates directory.
    /// </summary>
    /// 
    /// <returns>The templates directory.</returns>
    public DirectoryInfo? TemplatesDirectory
    {
        get
        {
            return this.templatesDir;
        }
    }

    /// <summary>
    /// Checks if the installation is actually a development build.
    /// <summary>
    /// 
    /// <returns>
    /// <c>true</c> if this installation represents a development
    /// build, otherwise <c>false</c>.
    /// </returns>
    public bool IsDevelopmentBuild
    {
        get
        {
            return this.devBuild;
        }
    }

    /// <summary>
    /// Produces a <c>string</c> describing this instance.
    /// 
    /// @return A <c>string</c> describing this instance.
    /// </summary>
    public override string ToString()
    {
        StringBuilder sb = new StringBuilder();

        sb.AppendLine();
        sb.AppendLine("--------------------------------------------------");
        sb.AppendLine("installDirectory   : " + this.InstallDirectory);
        sb.AppendLine("configDirectory    : " + this.ConfigDirectory);
        sb.AppendLine("supportDirectory   : " + this.SupportDirectory);
        sb.AppendLine("resourceDirectory  : " + this.ResourceDirectory);
        sb.AppendLine("templatesDirectory : " + this.TemplatesDirectory);
        sb.AppendLine("developmentBuild   : " + this.IsDevelopmentBuild);

        return sb.ToString();
    }

    /// <summary>
    /// Checks if the specified path is a directory.
    /// </summary>
    /// <param name="path">The path to check</param>
    /// <returns><c>true</c> if it is a directory, otherwise <c>false</c></returns>
    private static bool IsDirectory(string path)
    {
        FileAttributes attrs = File.GetAttributes(path);
        return (attrs & FileAttributes.Directory) == FileAttributes.Directory;
    }

    /// <summary>
    /// Finds the install directories and returns the <c>InstallLocations</c>
    /// instance describing those locations.
    ///
    /// @return The <c>InstallLocations</c> instance describing the install
    ///         locations.
    /// </summary>
    public static InstallLocations? FindLocations()
    {
        DirectoryInfo? installDir = null;
        DirectoryInfo? configDir = null;
        DirectoryInfo? resourceDir = null;
        DirectoryInfo? supportDir = null;
        DirectoryInfo? templatesDir = null;

        string defaultInstallPath;
        string? defaultConfigPath = null;
        string defaultSupportPath;

        switch (Environment.OSVersion.Platform)
        {
            case PlatformID.Win32NT:
                defaultInstallPath = "C:\\Program Files\\Senzing\\er";
                defaultSupportPath = "C:\\Program Files\\Senzing\\er\\data";
                break;
            case PlatformID.MacOSX:
                defaultInstallPath = "/opt/senzing/er";
                defaultSupportPath = "/opt/senzing/er/data";
                break;
            case PlatformID.Unix:
                defaultInstallPath = "/opt/senzing/er";
                defaultConfigPath = "/etc/opt/senzing";
                defaultSupportPath = "/opt/senzing/data";
                break;
            default:
                throw new NotSupportedException(
                    "Unsupported Operating System: "
                    + Environment.OSVersion.Platform);
        }

        // check for senzing system properties
        string? installPath = Environment.GetEnvironmentVariable(
            "SENZING_DIR");
        string? configPath = Environment.GetEnvironmentVariable(
            "SENZING_ETC_DIR");
        string? supportPath = Environment.GetEnvironmentVariable(
            "SENZING_DATA_DIR");
        string? resourcePath = Environment.GetEnvironmentVariable(
            "SENZING_RESOURCE_DIR");

        // normalize empty strings as null
        if (installPath != null && installPath.Trim().Length == 0)
        {
            installPath = null;
        }
        if (configPath != null && configPath.Trim().Length == 0)
        {
            configPath = null;
        }
        if (supportPath != null && supportPath.Trim().Length == 0)
        {
            supportPath = null;
        }
        if (resourcePath != null && resourcePath.Trim().Length == 0)
        {
            resourcePath = null;
        }

        // check the senzing directory
        installDir = new DirectoryInfo(
            installPath == null ? defaultInstallPath : installPath);
        if (!installDir.Exists)
        {
            Console.WriteLine("Could not find Senzing installation directory:");
            Console.WriteLine("     " + installDir);
            Console.WriteLine();
            if (installPath != null)
            {
                Console.WriteLine(
                    "Check the SENZING_DIR environment variable.");
            }
            else
            {
                Console.WriteLine(
                    "Use the SENZING_DIR environment variable to specify a path");
            }

            return null;
        }

        // normalize the senzing directory
        string installDirName = installDir.Name;
        if (installDir.Exists && IsDirectory(installDir.FullName)
            && !installDirName.Equals("er", OrdinalIgnoreCase)
            && installDirName.Equals("senzing", OrdinalIgnoreCase))
        {
            // for windows or linux allow the "Senzing" dir as well
            installDir = new DirectoryInfo(Path.Combine(installDir.FullName, "er"));
        }

        if (!IsDirectory(installDir.FullName))
        {
            Console.Error.WriteLine("Senzing installation directory appears invalid:");
            Console.Error.WriteLine("     " + installDir);
            Console.Error.WriteLine();
            if (installPath != null)
            {
                Console.Error.WriteLine(
                    "Check the SENZING_DIR environment variable.");
            }
            else
            {
                Console.Error.WriteLine(
                    "Use the SENZING_DIR environment variable to specify a path");
            }

            return null;
        }

        // check if an explicit support path has been specified
        if (supportPath == null || supportPath.Trim().Length == 0)
        {
            // check if using a dev build
            if ("dist".Equals(installDir.Name, OrdinalIgnoreCase))
            {
                // use the "data" sub-directory of the dev build
                supportDir = new DirectoryInfo(
                    Path.Combine(installDir.FullName, "data"));
            }
            else
            {
                switch (Environment.OSVersion.Platform)
                {
                    case PlatformID.Win32NT:
                        defaultSupportPath = Path.Combine(installDir.FullName, "data");
                        break;
                    case PlatformID.MacOSX:
                        defaultSupportPath = Path.Combine(installDir.FullName, "data");
                        break;
                    case PlatformID.Unix:
                        break;
                    default:
                        throw new NotSupportedException(
                             "Unsupported Operating System: "
                             + Environment.OSVersion.Platform);
                }

                // no explicit path, try the default support path
                supportDir = new DirectoryInfo(defaultSupportPath);
            }

        }
        else
        {
            // use the specified explicit path
            supportDir = new DirectoryInfo(supportPath);
        }

        if (!supportDir.Exists)
        {
            Console.Error.WriteLine("The support directory does not exist:");
            Console.Error.WriteLine("         " + supportDir);
            if (supportPath != null)
            {
                Console.Error.WriteLine(
                    "Check the SENZING_DATA_DIR environment variable.");
            }
            else
            {
                Console.Error.WriteLine(
                    "Use the SENZING_DATA_DIR environment variable to specify a path");
            }

            throw new InvalidOperationException(
                "The support directory does not exist: " + supportDir);
        }

        if (!IsDirectory(supportDir.FullName))
        {
            Console.Error.WriteLine("The support directory is invalid:");
            Console.Error.WriteLine("         " + supportDir);
            if (supportPath != null)
            {
                Console.Error.WriteLine(
                    "Check the SENZING_DATA_DIR environment variable.");
            }
            else
            {
                Console.Error.WriteLine(
                    "Use the SENZING_DATA_DIR environment variable to specify a path");
            }
            throw new InvalidOperationException(
                "The support directory is invalid: " + supportDir);

        }

        // check the config path
        if (configPath != null)
        {
            configDir = new DirectoryInfo(configPath);
        }

        // check for a dev build installation
        if (configDir == null && installDir != null && "dist".Equals(installDir.Name, OrdinalIgnoreCase))
        {
            configDir = new DirectoryInfo(Path.Combine(installDir.FullName, "data"));
        }

        // if still null and there is a default, then use it
        if (configDir == null && defaultConfigPath != null)
        {
            configDir = new DirectoryInfo(defaultConfigPath);
            if (!configDir.Exists)
            {
                configDir = null;
            }
        }

        // if still null, try to use the install's etc directory
        if (configDir == null && installDir != null)
        {
            configDir = new DirectoryInfo(Path.Combine(installDir.FullName, "etc"));
            if (!configDir.Exists)
            {
                configDir = null;
            }
        }

        if (configPath != null && configDir != null && !configDir.Exists)
        {
            Console.Error.WriteLine(
                "The SENZING_ETC_DIR environment variable specifies a path that does not exist:");
            Console.Error.WriteLine(
                "         " + configPath);

            throw new InvalidOperationException(
                "Explicit config path does not exist: " + configPath);
        }
        if (configDir != null && configDir.Exists)
        {
            if (!IsDirectory(configDir.FullName))
            {
                Console.Error.WriteLine(
                    "The SENZING_ETC_DIR environment variable specifies a file, not a directory:");
                Console.Error.WriteLine(
                    "         " + configPath);

                throw new InvalidOperationException(
                    "Explicit config path is not directory: " + configPath);
            }

            String[] requiredFiles = ["cfgVariant.json"];
            List<string> missingFiles = new List<string>(requiredFiles.Length);

            foreach (string fileName in requiredFiles)
            {
                DirectoryInfo? configFile = new DirectoryInfo(
                    Path.Combine(configDir.FullName, fileName));
                DirectoryInfo? supportFile = new DirectoryInfo(
                    Path.Combine(supportDir.FullName, fileName));
                if (!configFile.Exists && !supportFile.Exists)
                {
                    missingFiles.Add(fileName);
                }
            }
            if (missingFiles.Count > 0 && configPath != null)
            {
                Console.Error.WriteLine(
                    "The SENZING_ETC_DIR environment variable specifies an invalid config directory:");
                foreach (string missing in missingFiles)
                {
                    Console.Error.WriteLine(
                        "         " + missing + " was not found");
                }
                throw new InvalidOperationException(
                    "Explicit config path missing required files: " + missingFiles);
            }
        }

        // now determine the resource path
        resourceDir = (resourcePath == null) ? null : new DirectoryInfo(resourcePath);
        if (resourceDir == null && installDir != null)
        {
            resourceDir = new DirectoryInfo(
                Path.Combine(installDir.FullName, "resources"));
            if (!resourceDir.Exists) resourceDir = null;
        }

        if (resourceDir != null && resourceDir.Exists
            && IsDirectory(resourceDir.FullName))
        {
            templatesDir = new DirectoryInfo(
                Path.Combine(resourceDir.FullName, "templates"));
        }

        if (resourcePath != null)
        {
            if (resourceDir != null && !resourceDir.Exists)
            {
                Console.Error.WriteLine(
                    "The SENZING_RESOURCE_DIR environment variable specifies a path that does not exist:");
                Console.Error.WriteLine(
                    "         " + resourcePath);

                throw new InvalidOperationException(
                    "Explicit resource path does not exist: " + resourcePath);
            }

            if (resourceDir == null || !IsDirectory(resourceDir.FullName)
                || templatesDir == null || !templatesDir.Exists
                || !IsDirectory(templatesDir.FullName))
            {
                Console.Error.WriteLine(
                    "The SENZING_RESOURCE_DIR environment variable specifies an invalid "
                        + "resource directory:");
                Console.Error.WriteLine("         " + resourcePath);

                throw new InvalidOperationException(
                    "Explicit resource path is not valid: " + resourcePath);
            }

        }
        else if (resourceDir == null || !resourceDir.Exists || !IsDirectory(resourceDir.FullName)
            || templatesDir == null || !templatesDir.Exists || !IsDirectory(templatesDir.FullName))
        {
            resourceDir = null;
            templatesDir = null;
        }

        // construct and initialize the result
        InstallLocations result = new InstallLocations();
        result.installDir = installDir;
        result.configDir = configDir;
        result.supportDir = supportDir;
        result.resourceDir = resourceDir;
        result.templatesDir = templatesDir;
        result.devBuild = (installDir != null) && ("dist".Equals(installDir.Name, OrdinalIgnoreCase));

        // return the result
        return result;
    }
}
